### PR TITLE
PCHR-2553: Fix editing own request by manager

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

Currently managers cannot edit their own requests. This PR fixes this issue.

## Technical Details

*File js/angular/src/leave-absences/shared/controllers/request.controller.js*

The issue was in the `setRoles()` function - it ignored the `vm.isSelfRecord` value. `isSelfRecord` is essential for **My Leave** page since on this page all users with any permissions make/edit leave requests for **themselves**.

This means simply adding a force check on the `isSelfRecord` solves the issue:
```js
function initRoles () {
  role = 'staff';

  if (vm.isSelfRecord) {
    return; // <-- force the role to be "staff" in case of a self record
  }

  // Then if not a self record, set roles basing on permissions.
  // (ex. on Manager Leave page the functionality is different to Managers and Admins)
  return checkPermissions(sharedSettings.permissions.admin.administer).then...

}
```

---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 1]
```diff
@@ -498,16 +498,20 @@ define([
     function initRoles () {
       role = 'staff';
 
+      if (vm.isSelfRecord) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: We should not really change the `role` based on `isSelfRecord`. Because they are mutually exclusive, in future we might have a requirement where the my leave page needs to have different functionality based on the role(staff or manager). So we should keep the role as it is, but add conditions based on `isSelfRecord` to achieve the same.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 2]
```diff
@@ -828,6 +828,38 @@
         });
       });
 
+      describe('user edits their own leave request popup', function () {
+        var leaveRequest;
+
+        ['staff', 'manager', 'admin'].forEach(function (permissionsRole) {
+          testRoleForSelfRecord(permissionsRole);
+        })
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: Semicolon missing? Is the linter working for you?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js (hunk 1 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Sure. I am following the similar logic as per Leave Calendar:

```js
bindings: {
  contactId: '<',
  roleOverride: '@?'
}
```
```js
function setUserRole () {
  if (vm.roleOverride) {
    return $q.resolve().then(function () {
      userRole = vm.roleOverride;
    });
  } else {
    return $q.all([
      checkPermissions(sharedSettings.permissions.admin.administer),
      checkPermissions(sharedSettings.permissions.ssp.manage)
    ])
    .then(function (results) {
      userRole = results[0] ? 'admin' : (results[1] ? 'manager' : 'staff');
    });
  }
}
```
And here is the config file for the **my-leave** module:
```js
.state('my-leave.calendar', {
  url: '/calendar',
  template: '<leave-calendar contact-id="myleave.contactId" role-override="staff"></leave-calendar>'
});
```
I actually wanted initially to override the role basing on the `$rootScope.section`.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 3]
```diff
@@ -498,7 +498,7 @@ define([
     function initRoles () {
       role = 'staff';
 
-      if (vm.isSelfRecord) {
+      if ($rootScope.section === 'my-leave') {
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: is it possible to see my leave requests in a different section? can they be seen on manager leave? contact page?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 4]
```diff
@@ -376,7 +376,7 @@
 
                 leaveRequest.contact_id = CRM.vars.leaveAndAbsences.contactId.toString();
 
-                initTestController({ isSelfRecord: true, leaveRequest: leaveRequest });
+                initTestController({ leaveRequest: leaveRequest, isSelfRecord: true });
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: why did you remove `isSelfRecord: true` from other tests but left this one?

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: The implementation of `isSelfRecord` was changed and now is used only in one place in the code:

```js
function changeStatusBeforeSave () {
  if (vm.isSelfRecord) {
    vm.request.status_id = vm.requestStatuses[sharedSettings.statusNames.awaitingApproval].value;
  } else if (vm.canManage) {
    vm.request.status_id = vm.newStatusOnSave || vm.request.status_id;
  }
}
```
and exactly this place is tested: 

```js
it('status changes to waiting approval before calling API', function () {
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js (hunk 3 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Yes, this component is used in Absence Tab:

image
